docs/extras/casts.md

```diff
@@ -0,0 +1,201 @@
+
+```lean
+import data.complex.basic
+
+def from_R_to_C (r : ℝ) : ℂ := r
+
+#print from_R_to_C 
+-- def from_R_to_C : ℝ → ℂ := λ (r : ℝ), ↑r
+```
+
+Looking at the definition of the function, we see that Lean is using `↑` to mean "use a coercion".
+
+# A subtlety with automatic coercions
+
+Before we go on, let me get this potentially confusing issue out of the way. Because there's a coercion from `ℤ` to `ℝ` one can just write `a : ℝ` if `a` is an integer, and Lean knows you mean the corresponding real number. Watch out for this gotcha though:
+
+```lean
+variables (a b : ℤ) 
+
+#check (a + b : ℝ) -- ↑a + ↑b : ℝ
+#check ((a + b : ℤ) : ℝ) -- ↑(a + b) : ℝ
+```
+
+I was initially surprised that these two terms didn't evaluate to exactly the same thing. What is happening here is that for the first term, Lean knows that the type of `a + b` is supposed to be `ℝ`, so it then decides that the addition we want must be `real.add`, so it then decides that we must want `a` and `b` to be reals, giving us `↑a + ↑b`. In the second term, we explicitly say that we want `a + b` to be treated like an integer, and so Lean does the addition first.
+
+# How do I know which coercion the arrow means?
+
+If the tricks at the top work for you, you might not even have to worry about exactly what `↑a` means. But if you do not understand what your goal says, because you have lost track of whether `↑a` means the rational number `a`, or the real number `a` or the complex number `a`, and you do need to know, then try writing `set_option pp.all true` before your theorem and then taking another look. For example, what was `↑(a + b) = ↑a + ↑b` above might now become around 25 lines of output, starting with
+
+```
+⊢ @eq.{1} real
+    (@coe.{1 1} int real
+       (@coe_to_lift.{1 1} int real
+          (@coe_base.{1 1} int real
+	  ...
+```
+
+and the very first line tells you that the goal is an equality between two real numbers (and the second line indicates that the left hand side of this equality is a coercion from the integers).
+
+# "Special" coercions.
+
+The coercion from `ℕ` to `ℤ` is not defined as the one coming from the fact that `ℤ` is a ring; it is more computationally efficient to use the constructor. The two functions are equal, but this is a theorem and not a definition. The coercion from `ℚ` to `ℝ` is noncomputable (although it doesn't have to be), and the coercion from `ℝ` to `ℂ` is defined by hand rather than being part of a general scheme of maps from `ℝ` to lots of places. 
+
+# Avoiding `simp`.
+
+If you do want to prove "trivial in maths2 goals "by hand" for some reason, then you can write `set_option trace.simplify.rewrite true` before your theorem, see what `simp` is doing, and then mimic it.
```

Comment:
```suggestion
If you do want to prove "trivial in maths goals "by hand" for some reason, then you can write `set_option trace.simplify.rewrite true` before your theorem, see what `simp` is doing, and then mimic it.
```

Reply:
Alternatively, you could mentions `squeeze_simp`.

Reply:
```suggestion
If you do want to prove "trivial in maths" goals "by hand" for some reason, then you can write `set_option trace.simplify.rewrite true` before your theorem, see what `simp` is doing, and then mimic it.
```